sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+
+    @dpex.kernel
+    def copyto_kernel(X, Y):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        for feature_idx in range(n_features):
+            Y[feature_idx, sample_idx] = X[feature_idx, sample_idx]
+
+    return copyto_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_broadcast_division_kernel(n_samples, n_features, work_group_size):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+
+    @dpex.kernel
+    def broadcast_division(X, v):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        divisor = v[sample_idx]
+
+        for feature_idx in range(n_features):
+            X[feature_idx, sample_idx] = X[feature_idx, sample_idx] / divisor
+
+    return broadcast_division[global_size, work_group_size]
+
+
+@lru_cache
+def make_center_shift_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def center_shift(previous_center, center, center_shift):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        tmp = zero
+
+        for feature_idx in range(n_features):
+            center_diff = (
+                previous_center[feature_idx, sample_idx]
+                - center[feature_idx, sample_idx]
+            )
+            tmp += center_diff * center_diff
+
+        center_shift[sample_idx] = tmp
+
+    return center_shift[global_size, work_group_size]
+
+
+@lru_cache
+def make_half_l2_norm_dim0_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def half_l2_norm(X, result):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        l2_norm = zero
+
+        for feature_idx in range(n_features):
+            item = X[feature_idx, sample_idx]
+            l2_norm += item * item
+
+        result[sample_idx] = l2_norm / 2
+
+    return half_l2_norm[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_1dim_kernel(n_samples, work_group_size, device, dtype):
+    local_n_iterations = math.floor(math.log2(work_group_size))
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group_id = dpex.get_group_id(0)
+        local_work_id = dpex.get_local_id(0)
+        first_work_id = local_work_id == 0
+
+        n_samples = v.shape[0]
+
+        shm = dpex.local.array(work_group_size, dtype=dtype)
+
+        first_sample_idx = group_id * work_group_size * 2
+        augend_idx = first_sample_idx + local_work_id
+        addend_idx = first_sample_idx + work_group_size + local_work_id
```

Comment:
Here [augend](https://www.merriam-webster.com/dictionary/augend) and [addend](https://www.merriam-webster.com/dictionary/addend) are the two values that are going to be summed together by this work item during the current iteration.

Reply:
Regarding the unittests, I agree it could be useful for iterating on specific kernels, future PR addressing kernel changes could add a unittest targetting their specific change.